posts/2020-03-12-Rust-1.42.md

```diff
@@ -0,0 +1,175 @@
+```
+
+If you're interested in learning more, we published [a post on the Inside Rust blog](https://blog.rust-lang.org/inside-rust/2020/03/04/recent-future-pattern-matching-improvements.html) discussing these changes as well as more improvements to pattern matching that we may bring to stable in the future!
+
+
+### [`matches!`]
+
+This release of Rust stabilizes a new macro, [`matches!`](https://doc.rust-lang.org/nightly/std/macro.matches.html). This macro accepts an expression and a pattern, and returns true if the pattern matches the expression. In other words:
+
+```rust
+// with the match keyword
+match self.partial_cmp(other) {
+    Some(Less) => true,
+    _ => false,
+}
+
+// with the matches! macro
+matches!(self.partial_cmp(other), Some(Less))
+```
+
+You can also use features like `|` patterns and `if` clauses:
+
+```rust
+let foo = 'f';
+assert!(matches!(foo, 'A'..='Z' | 'a'..='z'));
+
+let bar = Some(4);
+assert!(matches!(bar, Some(x) if x > 2));
+```
+
+### `use proc_macro::TokenStream;` now works
+
+In Rust 2018, we [removed the need for `extern crate`](https://doc.rust-lang.org/stable/edition-guide/rust-2018/module-system/path-clarity.html#no-more-extern-crate). But procedural macros were a bit special, and so when you were writing a procedural macro, you still needed to say `extern crate proc_macro;`.
+
+In this release, [you no longer need this line when working with the 2018 edition; you can use `use` like any other crate][cargo/7700]. Given that most projects will already have a line similar to `use proc_macro::TokenStream;`, this change will mean that you can delete the `extern crate proc_macro;` line and your code will still work. This change is small, but brings procedural macros closer to regular code.
+
+### Libraries
+
+- [`iter::Empty<T>` now implements `Send` and `Sync` for any `T`.][68348]
+- [`Pin::{map_unchecked, map_unchecked_mut}` no longer require the return type
+   to implement `Sized`.][67935]
+- [`io::Cursor` now implements `PartialEq` and `Eq`.][67233]
+- [`Layout::new` is now `const`.][66254]
+
+### Stabilized APIs
+
+- [`CondVar::wait_while`] & [`CondVar::wait_timeout_while`]
+- [`DebugMap::key`] & [`DebugMap::value`]
+- [`ManuallyDrop::take`]
+- [`ptr::slice_from_raw_parts_mut`] & [`ptr::slice_from_raw_parts`]
+
+### Other changes
+
+[relnotes-cargo]: https://github.com/rust-lang/cargo/blob/master/CHANGELOG.md#cargo-142-2020-03-12
+[relnotes-clippy]: TODO
+
+There are other changes in the Rust 1.42.0 release: check out what changed in [Rust][notes], [Cargo][relnotes-cargo], and [Clippy][relnotes-clippy].
+
+
+### Compatibility Notes
+
+We have two notable compatibility notes this release: a deprecation in the standard library, and a demotion of 32-bit Apple targets to Tier 3.
+
+#### Error::Description is deprecated
+
+Sometimes, mistakes are made. The `Error::description` method is now considered to be one of those mistakes. The problem is with its type signature:
+
+    fn description(&self) -> &str {
```

Comment:
Shouldn't this be wrapped in a code fence?

Reply:
Markdown also treats pieces of text indented with four spaces as a code block.

Reply:
Does MD understand this as Rust code though like this?

Reply:
@pietroalbini Does it? I know that is a variant in markdown, but I don't know if it's available in commonmark. I only know of stackoverflow that supports that.

Reply:
> Does MD understand this as Rust code though like this?

That depends on the implementation, if it auto-detects languages or not.

> I know that is a variant in markdown, but I don't know if it's available in commonmark.

Yep it's standard: https://spec.commonmark.org/0.29/#indented-code-block

Reply:
> That depends on the implementation, if it auto-detects languages or not.

Probably best not to risk it. :)

Reply:
highlightjs is usually pretty good at detection. I have no strong opinion, I just forgot the indented code blocks were standard.